### PR TITLE
Use wrapper for API

### PR DESCRIPTION
**Pull request by wchill**
### Short Description:

HUGE change. Wrap PGoApi so that we have a stable interface for the bot to use. Wrapper maintains state and should automatically cache API responses, marking them dirty when a RPC call causes state to be invalidated.

Wrapper also parses API responses to put data into objects. For now, these objects don't do anything fancy, and they can probably be better organized, but that can be fixed later as long as things that the bot is accessing don't change.
### Changes:
- Update PGoApi to commit `e7c395523e25e357215bb16ac243103f846ddc5b`
- Wrap PGoApi to make it easier to keep up with changes in the protobufs
- All API calls now go through `self.api_wrapper` to signify that they are using the wrapper
- API wrapper caches API calls and methods are called only if they either (a) change the game state or (b) have been invalidated by a different API call. If all methods in a JSON-RPC call are cached, no request is performed to the servers.
- API wrapper now maintains internal state instead of the bot.
- API wrapper converts API responses to objects based on the JSON-RPC methods called. Sensible defaults are provided for missing values.
- API wrapper automatically handles status code 52 (requesting too fast), Unknown PokeStop error (not logged in, issue #60), "GET_INVENTORY" key error (PGoApi returns False, issue #109), and malformed responses caused by unstable servers with an automatic sleep-retry loop.
- Provide type hinting for various functions.

@OpenPoGo/maintainers

### CODE REVIEW

**Comment by jtdroste** [comment]
Holy crap....so excited...

File: api/__init__.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        return 0
+
+    def create_request(self):
+        return self._api.create_request()
+
+    def call(self, ignore_expiration=False, ignore_cache=False):
+        methods, method_keys, self._pending_calls, self._pending_calls_keys = self._pending_calls, self._pending_calls_keys, {}, []
+        if self.get_expiration_time() < 60 and ignore_expiration is False:
+            print("[API] Token has expired, attempting to log back in...")
+            while self.login() is False:
```
Comment (jtdroste) [inline api/__init__.py]: I would have a max retry rather than constantly attempt to login

**Comment by wchill** [comment]
I'm gonna need a LOT of help for the various calls to the API. A lot of stuff needs to be moved over to using response objects.

**Comment by jtdroste** [comment]
Okay dokie - I'll look at helping out later tonight!  Can you make a checklist or something of things to do

**Comment by wchill** [comment]
```
✘ wchill@1337BookPro  ~/Projects/OpenPoGoBot/pokemongo_bot   feature/api-wrapper ●  grep -rnw -e "api" .
./__init__.py:24:from api import PoGoApi
./__init__.py:38:        self.api = None
./__init__.py:198:        # instantiate api
./__init__.py:199:        self.api = PoGoApi(provider=self.config.auth_service, username=self.config.username, password=self.config.password)
./__init__.py:200:        #self.api = PGoApi()
./__init__.py:205:        #while not self.api.login(self.config.auth_service, self.config.username, self.config.password):
./__init__.py:206:        while not self.api.login():
./__init__.py:215:        self.api.get_player().get_inventory()
./__init__.py:217:        response_dict = self.api.call()
./__init__.py:266:        self.api.recycle_inventory_item(item_id=item_id, count=count)
./__init__.py:267:        self.api.call()
./__init__.py:270:        self.api.get_inventory()
./__init__.py:271:        response = self.api.call()
./__init__.py:280:        self.api.get_player().get_inventory()
./__init__.py:282:        result = self.api.call()
./__init__.py:310:                    self.api.set_position(*self.position)
./__init__.py:331:        self.api.set_position(*self.position)
./__init__.py:345:        self.api.get_player()
./__init__.py:346:        self.api.get_hatched_eggs()
./__init__.py:347:        self.api.get_inventory()
./__init__.py:348:        self.api.check_awarded_badges()
./__init__.py:349:        self.api.call()
./__init__.py:352:        self.api.get_inventory()
./__init__.py:353:        response_dict = self.api.call()
./__init__.py:377:        self.api.get_inventory()
./__init__.py:378:        response_dict = self.api.call()
./cell_workers/initial_transfer_worker.py:11:        self.api = bot.api
./cell_workers/initial_transfer_worker.py:43:                    self.api.release_pokemon(pokemon_id=pokemon_groups[group_id][group_cp[i]])
./cell_workers/initial_transfer_worker.py:46:                    # response_dict = self.api.call()
./cell_workers/initial_transfer_worker.py:47:                    self.api.call()
./cell_workers/initial_transfer_worker.py:55:        self.api.get_player().get_inventory()
./cell_workers/initial_transfer_worker.py:56:        response_dict = self.api.call()
./cell_workers/pokemon_catch_worker.py:37:        self.api = bot.api
./cell_workers/pokemon_catch_worker.py:50:        self.api.catch_pokemon(
./cell_workers/pokemon_catch_worker.py:58:        response_dict = self.api.call()
./cell_workers/pokemon_catch_worker.py:97:        self.api.encounter(encounter_id=encounter_id,
./cell_workers/pokemon_catch_worker.py:101:        response_dict = self.api.call()
./cell_workers/pokemon_catch_worker.py:161:        self.api.get_inventory()
./cell_workers/pokemon_catch_worker.py:162:        response_dict = self.api.call()
./cell_workers/pokemon_catch_worker.py:184:        self.api.release_pokemon(pokemon_id=pid)
./cell_workers/pokemon_catch_worker.py:186:        # response_dict = self.api.call()
./cell_workers/pokemon_catch_worker.py:187:        self.api.call()
./cell_workers/pokemon_catch_worker.py:190:        self.api.get_inventory()
./cell_workers/pokemon_catch_worker.py:191:        response_dict = self.api.call()
./cell_workers/recycle_items_worker.py:9:        self.api = bot.api
./cell_workers/seen_fort_worker.py:16:        self.api = bot.api
./cell_workers/seen_fort_worker.py:27:        self.api.fort_search(fort_id=self.fort["id"],
./cell_workers/seen_fort_worker.py:32:        response_dict = self.api.call()
./cell_workers/walk_towards_fort_worker.py:11:        self.api = bot.api
./cell_workers/walk_towards_fort_worker.py:37:                self.api.set_position(*position)
./cell_workers/walk_towards_fort_worker.py:38:            self.api.player_update(latitude=lat, longitude=lng)
./cell_workers/walk_towards_fort_worker.py:42:        self.api.fort_details(fort_id=fort_id,
./cell_workers/walk_towards_fort_worker.py:45:        response_dict = self.api.call()
./stepper.py:19:        self.api = bot.api
./stepper.py:35:            position_lat, position_lng, _ = self.api.get_position()
./stepper.py:38:        self.api.set_position(*position)
./stepper.py:43:        position_lat, position_lng, _ = self.api.get_position()
./stepper.py:53:                position_lat, position_lng, _ = self.api.get_position()
./stepper.py:56:                self.api.set_position(c_lat, c_long, alt)
./stepper.py:61:                position_lat, position_lng, _ = self.api.get_position()
./stepper.py:68:        position_lat, position_lng, _ = self.api.get_position()
./stepper.py:85:        self.api.get_map_objects(latitude=f2i(lat),
./stepper.py:90:        response_dict = self.api.call()
```

Most of these will probably need to be changed.

**Comment by wchill** [comment]
Ready for feedback.

**Comment by MaxLeiter** [comment]
@jtdroste: think you could do a close review/test of this at some point?

**Comment by jtdroste** [comment]
I didn't run into any issues running it for a bit.

Looks like it's failing pep8 - btw.  I'll look at the code and do a review later

**Comment by wchill** [comment]
Yeah, I didn't thoroughly pylint it because I figured there would be some code that would need fixing anyway. No point pylinting it until ready to merge.

**Comment by meyer9** [comment]
:+1:

[Approved with PullApprove](https://pullapprove.com/OpenPoGo/OpenPoGoBot/pull-request/112/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=OpenPoGo/OpenPoGoBot)

**Comment by MaxLeiter** [comment]
:+1: awesome work!

[Approved with PullApprove](https://pullapprove.com/OpenPoGo/OpenPoGoBot/pull-request/112/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=OpenPoGo/OpenPoGoBot)

**Comment by wchill** [comment]
Squashing before merging.

**Comment by MaxLeiter** [comment]
:tada: LGTM

**Pull request merged**